src/libpriv/rpmostree-core.c

```diff
@@ -659,6 +659,37 @@ rpmostree_context_setup (RpmOstreeContext    *self,
   return TRUE;
 }
 
+/* For both server-side composes and client-side layering, we add the
+ * rpmmd repo information, since it's very useful for determining
+ * state.  See also:
+ *
+ * - https://github.com/projectatomic/rpm-ostree/issues/774
+ * - The repo_metadata_for_package() function in rpmostree-unpacker.c
+
+ * This appends an `rpmostree.rpmmd-repos` key of type aa{sv} to @builder, which should be of type a{sv}
+ */
+static void
+add_rpmmd_repo_commit_metadata (GVariantBuilder   *builder,
+                                DnfContext        *dnfctx,
+                                RpmOstreeTreespec *spec)
+{
+  g_auto(GVariantBuilder) repo_list_builder;
+  g_variant_builder_init (&repo_list_builder, (GVariantType*)"aa{sv}");
+  g_autoptr(GPtrArray) repos = get_enabled_rpmmd_repos (dnfctx, DNF_REPO_ENABLED_PACKAGES);
```

Comment:
This will not work in `--cache-only` mode though, right? We specifically disable all repos there. I guess we could iterate through the disabled repos anyway and pick up the timestamp from the last time it was updated. But then again, one might've done `refresh-md` separately anyway so it wouldn't be a good representation of the state of the repo at the time the package was fetched.

Hmm, it seems like we would want instead to store that information as part of unpacking, along with the repo id? And then here, just collect those timestamps into a set?

Reply:
Yeah, that's a good point, though also a huge can of worms 😄  But I agree that if we're going to try to have a "show how up to date you are" feature it's problematic if it fails in the corner cases like mixing in caching operations, since that's when one most needs it.

I think you're right that the import is currently the "ground truth" and the data needs to live there, but what hurts my head is trying to figure out how to deal with that set.  It's going to be very common for packages not to update even if the repo timestamp changes.  Do we update the commit?  Or (probably) pick the largest timestamp from the set per repo and use that?

Doing it this way also has a backcompat issue in that we need to deal with systems that predate this change.  But that's doable.

Reply:
>  It's going to be very common for packages not to update even if the repo timestamp changes. Do we update the commit? Or (probably) pick the largest timestamp from the set per repo and use that?

Yeah, definitely tricky to think about! :)

Though high-level, going back to https://github.com/projectatomic/rpm-ostree/issues/774, it seems like the proper way to answer "Is my system up to date?" is to implement something like https://github.com/projectatomic/rpm-ostree/issues/247, right? This is something I've thought about a lot recently (and would love to tackle next! :)). Basically, I really *really* want to be able to type `rpm-ostree status` on my Atomic Workstation and have it tell me that there is a pending commit/pending packages waiting and ideally tell me exactly what the pending diff is (maybe hidden under `-v`).

For sure, tracking timestamps helps too if e.g. auto-metadata fetching is disabled or if offline often. But it seems like doing what you mentioned (picking the largest timestamp from the set per repo) is probably "good enough" given that it will be properly solved by https://github.com/projectatomic/rpm-ostree/issues/247?

Reply:
Yes, I think automatic updates are going to obviate this type of thing for a lot of uses - but here's a good example of an issue it doesn't solve: how do we help users/admins deal with changing `/etc/yum.repos.d`?  I really think we need more visibility into that than we have today; probably show in bold if e.g. your booted tree's repos differ from the on-disk configured ones.

Particularly because of our anti-hysteresis model, one can have packages drop out if say a repo with `skip_if_unavailable=True` (like COPR) is temporarily unavailable.  (We really need to handle that better...`skip_if_unavailable` is basically designed around a traditional model where the packages will stay on your system by default)